transport/src/main/java/io/servicefabric/transport/protocol/Message.java

```diff
@@ -1,100 +1,124 @@
 package io.servicefabric.transport.protocol;
 
+import com.google.common.base.Preconditions;
+
 import javax.annotation.concurrent.Immutable;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
 
 /**
- * The Class Message introduce common protocol for transport to transport communication.
+ * The Class Message introduces generic protocol used for point to point communication by transport.
  */
 @Immutable
 public final class Message {
 
-	private final String qualifier;
+	private Map<String, String> headers;
 
-	private final String correlationId;
+	private Object data;
 
-	private final Object data;
+	/**
+	 * Instantiates empty message for deserialization purpose.
+	 */
+	Message() {
+	}
 
 	/**
-	 * Instantiates a new message with given qualifier. All other parameters set to null.
+	 * Instantiates a new message with the given data and without headers.
 	 */
-	public Message(String qualifier) {
-		this(qualifier, null);
+	public Message(Object data) {
+		setData(data);
+		this.headers = Collections.emptyMap();
 	}
 
 	/**
-	 * Instantiates new messages with given qualifier and data.
+	 * Instantiates a new message with the given headers and with empty data.
 	 */
-	public Message(String qualifier, Object data) {
-		this(qualifier, data, null);
+	public Message(Map<String, String> headers) {
+		setHeaders(headers);
 	}
 
 	/**
 	 * Instantiates new messages with given qualifier, data and correlationId.
 	 */
-	public Message(String qualifier, Object data, String correlationId) {
-		this.qualifier = qualifier;
+	public Message(Object data, Map<String, String> headers) {
+		setData(data);
+		setHeaders(headers);
+	}
+
+	/**
+	 * Instantiates new message with given qualifier and headers. Headers passed a sequence of key-value pairs.
+	 */
+	public Message(Object data, String... headers) {
+		Preconditions.checkArgument(headers.length % 2 == 0, "");
+		Map<String, String> headersMap = new HashMap<>(headers.length / 2);
+		for (int i = 0; i < headers.length; i += 2) {
+			headersMap.put(headers[i], headers[i + 1]);
+		}
+		setHeaders(headersMap);
+		setData(data);
+	}
+
+	void setData(Object data) {
 		this.data = data;
-		this.correlationId = correlationId;
+	}
+
+	void setHeaders(Map<String, String> headers) {
+		if (headers == null) {
+			this.headers = Collections.emptyMap();
```

Comment:
Let ```this.headers = Collections.emptyMap()``` be from very start, in initialization block. Otherwise if no one would call ```setHeaders``` this field will be null.